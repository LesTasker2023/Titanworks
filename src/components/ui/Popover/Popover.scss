@import '../../../styles/design-tokens';

.popover {
  // Content container
  &__content {
    z-index: 50;
    width: 18rem;
    border-radius: var(--radius);
    border: var(--border-width-thin) solid var(--border);
    background-color: var(--popover);
    padding: 1rem;
    color: var(--popover-foreground);
    box-shadow: var(--shadow-md);
    outline: none;
    transform-origin: var(--radix-popover-content-transform-origin);
    // Animation states
    &[data-state="open"] {
      animation:
        popover-fade-in 0.2s ease-out,
        popover-zoom-in 0.2s ease-out;
    }
    &[data-state="closed"] {
      animation:
        popover-fade-out 0.2s ease-out,
        popover-zoom-out 0.2s ease-out;
    }
    // Side-specific slide animations
    &[data-side="bottom"] {
      animation:
        popover-fade-in 0.2s ease-out,
        popover-zoom-in 0.2s ease-out,
        popover-slide-in-from-top 0.2s ease-out;
    }
    &[data-side="left"] {
      animation:
        popover-fade-in 0.2s ease-out,
        popover-zoom-in 0.2s ease-out,
        popover-slide-in-from-right 0.2s ease-out;
    }
    &[data-side="right"] {
      animation:
        popover-fade-in 0.2s ease-out,
        popover-zoom-in 0.2s ease-out,
        popover-slide-in-from-left 0.2s ease-out;
    }
    &[data-side="top"] {
      animation:
        popover-fade-in 0.2s ease-out,
        popover-zoom-in 0.2s ease-out,
        popover-slide-in-from-bottom 0.2s ease-out;
    }
    &[data-state="closed"][data-side="bottom"] {
      animation:
        popover-fade-out 0.2s ease-out,
        popover-zoom-out 0.2s ease-out,
        popover-slide-out-to-top 0.2s ease-out;
    }
    &[data-state="closed"][data-side="left"] {
      animation:
        popover-fade-out 0.2s ease-out,
        popover-zoom-out 0.2s ease-out,
        popover-slide-out-to-right 0.2s ease-out;
    }
    &[data-state="closed"][data-side="right"] {
      animation:
        popover-fade-out 0.2s ease-out,
        popover-zoom-out 0.2s ease-out,
        popover-slide-out-to-left 0.2s ease-out;
    }
    &[data-state="closed"][data-side="top"] {
      animation:
        popover-fade-out 0.2s ease-out,
        popover-zoom-out 0.2s ease-out,
        popover-slide-out-to-bottom 0.2s ease-out;
    }
  }
}
// Animations
@keyframes popover-fade-in {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
@keyframes popover-fade-out {
  from {
    opacity: 1;
  }
  to {
    opacity: 0;
  }
}
@keyframes popover-zoom-in {
  from {
    transform: scale(0.95);
  }
  to {
    transform: scale(1);
  }
}
@keyframes popover-zoom-out {
  from {
    transform: scale(1);
  }
  to {
    transform: scale(0.95);
  }
}
@keyframes popover-slide-in-from-top {
  from {
    transform: translateY(-0.5rem);
  }
  to {
    transform: translateY(0);
  }
}
@keyframes popover-slide-in-from-bottom {
  from {
    transform: translateY(0.5rem);
  }
  to {
    transform: translateY(0);
  }
}
@keyframes popover-slide-in-from-left {
  from {
    transform: translateX(-0.5rem);
  }
  to {
    transform: translateX(0);
  }
}
@keyframes popover-slide-in-from-right {
  from {
    transform: translateX(0.5rem);
  }
  to {
    transform: translateX(0);
  }
}
@keyframes popover-slide-out-to-top {
  from {
    transform: translateY(0);
  }
  to {
    transform: translateY(-0.5rem);
  }
}
@keyframes popover-slide-out-to-bottom {
  from {
    transform: translateY(0);
  }
  to {
    transform: translateY(0.5rem);
  }
}
@keyframes popover-slide-out-to-left {
  from {
    transform: translateX(0);
  }
  to {
    transform: translateX(-0.5rem);
  }
}
@keyframes popover-slide-out-to-right {
  from {
    transform: translateX(0);
  }
  to {
    transform: translateX(0.5rem);
  }
}
