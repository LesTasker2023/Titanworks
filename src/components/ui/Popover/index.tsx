export { default as PopoverDemo } from './demo';
export * from './Popover';
