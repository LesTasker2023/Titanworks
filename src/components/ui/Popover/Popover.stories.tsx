import type { Meta, StoryObj } from '@storybook/nextjs';
import { Button } from '../Button/Button';
import { Popover, PopoverContent, PopoverTrigger } from './Popover';

const meta: Meta<typeof Popover> = {
  title: 'UI/Popover',
  component: Popover,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {},
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: () => (
    <Popover>
      <PopoverTrigger asChild>
        <Button variant="outline">Open Popover</Button>
      </PopoverTrigger>
      <PopoverContent>
        <div className="space-y-2">
          <h4 className="font-medium leading-none">Dimensions</h4>
          <p className="text-sm text-muted-foreground">Set the dimensions for the layer.</p>
        </div>
      </PopoverContent>
    </Popover>
  ),
};

export const WithCustomContent: Story = {
  render: () => (
    <Popover>
      <PopoverTrigger asChild>
        <Button variant="outline">Settings</Button>
      </PopoverTrigger>
      <PopoverContent className="w-80">
        <div className="space-y-4">
          <div className="space-y-2">
            <h4 className="font-medium leading-none">Account Settings</h4>
            <p className="text-sm text-muted-foreground">
              Manage your account settings and preferences.
            </p>
          </div>
          <div className="space-y-2">
            <label htmlFor="username" className="text-sm font-medium">
              Username
            </label>
            <input
              id="username"
              type="text"
              placeholder="Enter username"
              className="w-full px-3 py-2 border rounded-md"
            />
          </div>
          <div className="flex justify-end">
            <Button size="sm">Save Changes</Button>
          </div>
        </div>
      </PopoverContent>
    </Popover>
  ),
};
