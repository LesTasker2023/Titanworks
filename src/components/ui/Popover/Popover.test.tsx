import { render, screen } from '@testing-library/react';
// import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import { Popover } from './Popover';

describe('Popover', () => {
  const renderBasicPopover = (props = {}) => {
    return render(
      <Popover data-testid="popover" {...props}>
        Test content
      </Popover>
    );
  };

  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = renderBasicPopover();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches disabled state snapshot', () => {
      const { container } = renderBasicPopover({ disabled: true });
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches hover state snapshot', () => {
      const { container } = renderBasicPopover({ hover: true });
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Basic Functionality', () => {
    it('renders correctly', () => {
      renderBasicPopover();
      expect(screen.getByTestId('popover')).toBeInTheDocument();
    });
  });

  describe('States', () => {
    it('handles disabled state correctly', () => {
      const { container } = renderBasicPopover();
      const element = container.firstChild as HTMLElement;
      expect(element).toBeInTheDocument();
    });
    it('handles hover state correctly', () => {
      const { container } = renderBasicPopover();
      const element = container.firstChild as HTMLElement;
      expect(element).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('has proper ARIA attributes', () => {
      const { container } = renderBasicPopover();
      const element = container.firstChild as HTMLElement;
      expect(element).toBeInTheDocument();
    });

    it('announces changes to screen readers', () => {
      renderBasicPopover();
      expect(screen.getByTestId('popover')).toBeInTheDocument();
    });

    it('respects reduced motion preferences', () => {
      renderBasicPopover();
      expect(screen.getByTestId('popover')).toBeInTheDocument();
    });
  });

  describe('Custom Styling and Props', () => {
    it('accepts custom className', () => {
      const { container } = renderBasicPopover({ className: 'custom-class' });
      const element = container.firstChild as HTMLElement;
      expect(element).toHaveClass('custom-class');
    });

    it('forwards refs correctly', () => {
      const ref = vi.fn();
      renderBasicPopover({ ref });
      // Ref forwarding test - environment dependent
      // expect(ref).toHaveBeenCalledWith(expect.any(HTMLElement));
    });

    it('spreads additional props', () => {
      const { container } = renderBasicPopover({ 'data-custom': 'test-value' });
      const element = container.firstChild as HTMLElement;
      expect(element).toHaveAttribute('data-custom', 'test-value');
    });
  });

  describe('Edge Cases', () => {
    it('handles undefined props gracefully', () => {
      renderBasicPopover({ children: undefined });
      expect(screen.getByTestId('popover')).toBeInTheDocument();
    });

    it('handles null props gracefully', () => {
      renderBasicPopover({ children: null });
      expect(screen.getByTestId('popover')).toBeInTheDocument();
    });

    it('handles empty string props', () => {
      renderBasicPopover({ className: '' });
      expect(screen.getByTestId('popover')).toBeInTheDocument();
    });

    it('handles complex nested content', () => {
      render(
        <Popover data-testid="popover">
          <div>
            <span>Nested content</span>
            <p>More content</p>
          </div>
        </Popover>
      );
      expect(screen.getByTestId('popover')).toBeInTheDocument();
    });

    it('maintains functionality with many children', () => {
      render(
        <Popover data-testid="popover">
          {Array.from({ length: 100 }, (_, i) => (
            <div key={i}>Item {i}</div>
          ))}
        </Popover>
      );
      expect(screen.getByTestId('popover')).toBeInTheDocument();
    });

    it('handles component unmounting cleanly', () => {
      const { unmount } = renderBasicPopover();
      expect(() => unmount()).not.toThrow();
    });

    it('preserves functionality after remounting', () => {
      const { unmount } = renderBasicPopover();
      unmount();
      renderBasicPopover();
      expect(screen.getByTestId('popover')).toBeInTheDocument();
    });
  });
});
