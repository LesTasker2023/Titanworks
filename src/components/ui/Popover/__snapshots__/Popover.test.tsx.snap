// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Popover > Snapshots > matches default snapshot 1`] = `
<button
  aria-controls="radix-«r0»"
  aria-expanded="false"
  aria-haspopup="dialog"
  data-state="closed"
  data-testid="popover-trigger"
  type="button"
>
  <div>
    Open Popover
  </div>
</button>
`;

exports[`Popover > Snapshots > matches disabled state snapshot 1`] = `
<button
  aria-controls="radix-«r1»"
  aria-expanded="false"
  aria-haspopup="dialog"
  data-state="closed"
  data-testid="popover-trigger"
  type="button"
>
  <div>
    Open Popover
  </div>
</button>
`;

exports[`Popover > Snapshots > matches hover state snapshot 1`] = `
<button
  aria-controls="radix-«r2»"
  aria-expanded="false"
  aria-haspopup="dialog"
  data-state="closed"
  data-testid="popover-trigger"
  type="button"
>
  <div>
    Open Popover
  </div>
</button>
`;
